### PR TITLE
Laboratory work #2, Viktoriya Aleksina - 23FPL2

### PR DESCRIPTION
**Pull request by AleksinaV**

### CODE REVIEW

File: lab_2_retrieval_w_bm25/start.py [hunk 1]
```diff
@@ -26,7 +27,8 @@ def main() -> None:
             documents.append(file.read())
     with open("assets/stopwords.txt", "r", encoding="utf-8") as file:
         stopwords = file.read().split("\n")
-    result = None
+    result = tokenize(documents[0])
```
[the remark this answers is not in the document]
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/start.py]: done

File: lab_2_retrieval_w_bm25/main.py [hunk 2]
```diff
@@ -33,6 +41,17 @@ def remove_stopwords(tokens: list[str], stopwords: list[str]) -> list[str] | Non
...
     In case of corrupt input arguments, None is returned.
     """
+    if (not tokens or not isinstance(tokens, list)
+            or not all(isinstance(i, str) for i in tokens)):
+        return None
+    if (not stopwords or not isinstance(stopwords, list)
+            or not all(isinstance(i, str) for i in stopwords)):
+        return None
+
+    for token in tokens.copy():
```
[the remark this answers is not in the document]
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py:52]: Is there something I need to fix? I don't see it(

File: lab_2_retrieval_w_bm25/main.py [hunk 3]
```diff
@@ -207,3 +364,15 @@ def calculate_spearman(rank: list[int], golden_rank: list[int]) -> float | None:
...
     """
+    is_not_correct = (not rank or not isinstance(rank, list) or
+                      not all(isinstance(i, int) for i in rank) or
+                      not golden_rank or not isinstance(golden_rank, list) or
+                      not all(isinstance(i, int) for i in golden_rank) or
+                      not len(rank) == len(golden_rank))
+    if is_not_correct:
+        return None
+
+    n = len(rank)
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:375]: `n` is a bad name

File: lab_2_retrieval_w_bm25/main.py [hunk 4]
```diff
@@ -207,3 +364,15 @@ def calculate_spearman(rank: list[int], golden_rank: list[int]) -> float | None:
...
+    return 1 - (6 * sum((index - golden_rank.index(number)) ** 2
+                                        for index, number in enumerate(rank)
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:377]: It is absolutelly unreadble, move the loop outside the return statement

File: lab_2_retrieval_w_bm25/main.py [hunk 5]
```diff
@@ -168,6 +275,34 @@ def calculate_bm25_with_cutoff(
...
+    is_not_correct = (not vocab or not isinstance(vocab, list) or
+                      not all(isinstance(i, str) for i in vocab) or
+                      not document or not isinstance(document, list) or
+                      not all(isinstance(i, str) for i in document) or
+                      not idf_document or not isinstance(idf_document, dict) or
+                      not all((isinstance(key, str) and isinstance(value, float)
+                               for key, value in idf_document.items())) or
+                      all(isinstance(i, dict) for i in idf_document.values()) or
+                      not alpha or not isinstance(alpha, float) or
+                      not k1 or not isinstance(k1, float) or not 1.2 <= k1 <= 2.0 or
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:287]: magic numbers, Where did you find it?

File: lab_2_retrieval_w_bm25/main.py [hunk 6]
```diff
@@ -138,17 +225,37 @@ def rank_documents(
...
 def calculate_bm25_with_cutoff(
-    vocab: list[str],
-    document: list[str],
-    idf_document: dict[str, float],
-    alpha: float,
-    k1: float = 1.5,
-    b: float = 0.75,
-    avg_doc_len: float | None = None,
-    doc_len: int | None = None,
+        vocab: list[str],
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:251]: revert

File: lab_2_retrieval_w_bm25/main.py [hunk 7]
```diff
@@ -138,17 +225,37 @@ def rank_documents(
...
+        return None
+
+    tokenized_query = tokenize(query)
+    if not tokenized_query:
+        return None
+    query_preprocess = remove_stopwords(tokenized_query, stopwords)
+    if not query_preprocess:
+        return None
+
+    return sorted([(indexes.index(document), sum(document[word] for word in document
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:245]: too hard to understand

File: lab_2_retrieval_w_bm25/main.py [hunk 8]
```diff
@@ -120,10 +181,36 @@ def calculate_bm25(
...
+        num_term_occur = document.count(term)
+        bm25_dict[term] = (idf_document[term] * ((num_term_occur * (k1 + 1)) /
+                                                 (num_term_occur + k1 *
+                                                  (1 - b + b * doc_len / avg_doc_len))))
+    return bm25_dict
 
 
 def rank_documents(
-    indexes: list[dict[str, float]], query: str, stopwords: list[str]
+        indexes: list[dict[str, float]], query: str, stopwords: list[str]
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:213]: revert

File: lab_2_retrieval_w_bm25/main.py [hunk 9]
```diff
@@ -120,10 +181,36 @@ def calculate_bm25(
...
     """
+    is_not_correct = (not vocab or not isinstance(vocab, list) or
+                      not all(isinstance(i, str) for i in vocab) or
+                      not document or not isinstance(document, list) or
+                      not all(isinstance(i, str) for i in document) or
+                      not idf_document or not isinstance(idf_document, dict) or
+                      not all((isinstance(key, str) and isinstance(value, float)
+                               for key, value in idf_document.items())) or
+                      all(isinstance(i, dict) for i in idf_document.values()) or
+                      not k1 or not isinstance(k1, float) or not 1.2 <= k1 <= 2.0 or
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:192]: .

File: lab_2_retrieval_w_bm25/main.py [hunk 10]
```diff
@@ -92,16 +143,26 @@ def calculate_tf_idf(tf: dict[str, float], idf: dict[str, float]) -> dict[str, f
...
 
 def calculate_bm25(
-    vocab: list[str],
-    document: list[str],
-    idf_document: dict[str, float],
-    k1: float = 1.5,
-    b: float = 0.75,
-    avg_doc_len: float | None = None,
-    doc_len: int | None = None,
+        vocab: list[str],
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:159]: .

File: lab_2_retrieval_w_bm25/main.py [hunk 11]
```diff
@@ -47,11 +67,18 @@ def build_vocabulary(documents: list[list[str]]) -> list[str] | None:
...
+    if is_not_correct:
+        return None
+
+    return list(sum(documents, []))
 
 
 def calculate_tf(vocab: list[str], document_tokens: list[str]) -> dict[str, float] | None:
     """
-    Calculate term frequency for the given tokens based on the vocabulary.
+     Calculate term frequency for the given tokens based on the vocabulary.
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/main.py:81]: .

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py:287]: In the documentation
image

File: lab_2_retrieval_w_bm25/main.py (hunk 8 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py:213]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 10 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py:159]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 9 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py:192]: What exactly should I fix here?

File: lab_2_retrieval_w_bm25/main.py (hunk 11 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py:81]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 7 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 5 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py]: But I deleted it anyway

File: lab_2_retrieval_w_bm25/main.py (hunk 4 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 3 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/main.py (hunk 6 above)
Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/main.py]: done

File: lab_2_retrieval_w_bm25/start.py [hunk 12]
```diff
@@ -1,7 +1,9 @@
 """
 Laboratory Work #2 starter
 """
-# pylint:disable=too-many-locals, unused-argument, unused-variable, too-many-branches, too-many-statements, duplicate-code
+# pylint:disable=too-many-locals, unused-argument, unused-variable, too-many-branches,
```
Comment (artyomtugaryov) [inline lab_2_retrieval_w_bm25/start.py:4]: ? revert

Comment (AleksinaV) [reply lab_2_retrieval_w_bm25/start.py]: done

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**